### ISSUE TITLE
Small genome matrices are hard to display

### ISSUE THREAD

**Issue by gtrichard** [higlass/higlass#593]
First of all, thank you for developing HiGlass, this is an awesome tool.

One of the issue we currently have is that when loading a small genome matrix, like _Drosophila melanogaster_, the Hi-C matrix is **very** small. More often than not, we have to carefully zoom in and try to find the matrix, and usually we just "loose" the matrix (zoomed out too much, or zoomed in but not in the correct place). It is the case both when displaying a Hi-C matrix on side panels or at the center (although it is easier to find the matrix on the side panels of a window).

Wouldn't it be possible to automatically resize the window coordinates to the actual beginning and end of a matrix when it's added to the center or lateral panels?

Best,
Gautier

**Comment by flekschas** [comment]
Thanks for reaching out! Do you have an example view config? Also, have you tried `Zoom to data extent` yet? I believe it should do exactly what you're asking for.

Apart from that the matrix should only be very small if your initial X and Y domain (defined in your view config) are very large compared to the extent of your data. Have you tried adjusting it to the extent of the Drosophila melanogaster genome?

 In terms of preventing extensive zoom out you can first set the `initialXDomain` and `initialYDomain` to your genome's extent and then set `zoomLimit: [0.5, null]` this should prevent you from zooming out too much but it still allows infinite zoom in.

**Comment by pkerpedjiev** [comment]
It sounds like Gautier is suggesting we do those things automatically :) I would tend to agree that it would be a nice thing.

If there is no data shown and a matrix is added, we should `Zoom to data extent` automatically. That's actually really easy to implement. I can try to add it to the next minor release.

@gtrichard in the meantime, using the `Zoom to data extent` option from the view config menu is probably your best bet.

**Comment by gtrichard** [comment]
Thank you! This is exactly what I was looking for!

Doing it automatically would nice for sure, especially for newcomers ;)

But once you see that button, it is pretty obvious actually, thank you again @flekschas  @pkerpedjiev

**Comment by alexpreynolds** [comment]
For those working with the view programmatically and who want to handle when the view is zoomed out too much, one approach I have found is to do the following:

1. Track the location via the API, e.g.:

```
this.hgView.api.on("location", (event) => { 
  this.updateViewerLocation(event);
});
```

2. Set up a "debounced" function to update the viewer location state only after the last event is sent, e.g. with a 500ms delay between subsequent location events:

```
updateViewerLocation = (event) => {
  if (!this.viewerLocationChangeEventTimer) {
    clearTimeout(this.viewerLocationChangeEventTimer);
    this.viewerLocationChangeEventTimer = setTimeout(() => {
      console.log("hgView location changed");
      this.testIfAtZoomExtentBounds(event);
      this.viewerLocationChangeEventTimer = null;
    }, 500);
  }
}
```

If events are handled too frequently, the browser performance seems to suffer.

3. In `testIfAtZoomExtentBounds()`, get the chromosome, start, and stop positions of the view, and test if they are equal to the bounds for the first and last chromosome of the chromosome sizes file for the specified genome assembly.

If the view location is at the genome's extents, then it is either showing all of the data or it is zoomed out past the genome's bounds (the location event only reports up to the bounds). 

I then use a `hgApi.zoomTo()` call to zoom to the bounds, minus ten bases on the last chromosome. 

This brings the view just into the genome's extents by a little bit, but given the size of the genome we are working with (human), ten bases isn't much. 

The `zoomTo()` call triggers another location update, but now the view's location is within the clipped bounds we set, so there is no subsequent `zoomTo()` call until the view is once again zoomed out past the genome extents.
